src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -146,6 +150,23 @@ private async Task ExecuteRequests(Hl7.Fhir.Model.Bundle responseBundle, Hl7.Fhi
                 if (request.Handler != null)
                 {
                     HttpContext httpContext = request.HttpContext;
+
+                    IFhirRequestContext originalFhirRequestContext = _fhirRequestContextAccessor.FhirRequestContext;
+
+                    request.RouteData.Values.TryGetValue(KnownActionParameterNames.ResourceType, out object resourceType);
+                    var newFhirRequestContext = new FhirRequestContext(
+                        httpContext.Request.Method,
+                        httpContext.Request.GetDisplayUrl(),
+                        originalFhirRequestContext.BaseUri.OriginalString,
+                        originalFhirRequestContext.CorrelationId,
+                        httpContext.Request.Headers,
+                        httpContext.Response.Headers,
+                        resourceType?.ToString())
+                    {
+                        Principal = originalFhirRequestContext.Principal,
+                    };
+                    _fhirRequestContextAccessor.FhirRequestContext = newFhirRequestContext;
```

Comment:
Do we not need to restore this context after it's done?

Reply:
I feel like there are a lot of conversions happening (hence more allocations and etc). And now reading through the code, mocking with the context and "simulating" the calls concerns me a bit also and i am not sure if we have enough test coverage in various areas we have to confidently say this works without regressions. Let's chat more on Monday.

Reply:
Because the accessor is backed by an asynclocal, the flow should only carry from outer async calls to inner async calls. Once you're out of the particular async call, it should revert to the value that it was.